core/modules/query/WhereClause.cc

```diff
@@ -107,122 +105,144 @@ void findColumnRefs(std::shared_ptr<BoolTerm> t, ColumnRef::Vector& vector) {
+        _rootOrTerm->renderTo(qt);
     }
 }
 
+
 std::shared_ptr<WhereClause> WhereClause::clone() const {
     // FIXME
     std::shared_ptr<WhereClause> newC = std::make_shared<WhereClause>(*this);
     // Shallow copy of expr list is okay.
-    if (_tree.get()) {
-        newC->_tree = _tree->copySyntax();
+    if (nullptr != _rootOrTerm) {
+        newC->_rootOrTerm = _rootOrTerm->copy();
     }
-    if (_restrs.get()) {
+    if (nullptr != _restrs) {
         newC->_restrs = std::make_shared<QsRestrictor::PtrVector>(*_restrs);
     }
     // For the other fields, default-copied versions are okay.
     return newC;
 
 }
 
+
 std::shared_ptr<WhereClause> WhereClause::copySyntax() {
     std::shared_ptr<WhereClause> newC = std::make_shared<WhereClause>(*this);
     // Shallow copy of expr list is okay.
-    if (_tree.get()) {
-        newC->_tree = _tree->copySyntax();
+    if (nullptr != _rootOrTerm) {
+        newC->_rootOrTerm = _rootOrTerm->copy();
     }
     // For the other fields, default-copied versions are okay.
     return newC;
 }
 
-void
-WhereClause::prependAndTerm(std::shared_ptr<BoolTerm> t) {
-    // Walk to AndTerm and prepend new BoolTerm in front of the
-    // list. If the new BoolTerm is an instance of AndTerm, prepend
-    // its terms rather than the AndTerm itself.
-    std::shared_ptr<BoolTerm> insertPos = skipTrivialOrTerms(_tree);
-
-    // FIXME: Should deal with case where AndTerm is not found.
-    AndTerm* rootAnd = dynamic_cast<AndTerm*>(insertPos.get());
-    if (!rootAnd) {
-        std::shared_ptr<AndTerm> a = std::make_shared<AndTerm>();
-        std::shared_ptr<BoolTerm> oldTree(_tree);
-        _tree = a;
-        if (oldTree.get()) { // Only add oldTree root if non-NULL
-            a->_terms.push_back(oldTree);
-        }
-        rootAnd = a.get();
 
+std::shared_ptr<AndTerm> WhereClause::_addRootAndTerm() {
+    if (nullptr == _rootOrTerm) {
+        _rootOrTerm = std::make_shared<OrTerm>();
+    } else if (_rootOrTerm->_terms.size() != 0) {
+        throw std::logic_error("Can not add root AND term."); // expected 0 or 1 items in _terms
     }
-    if (!rootAnd) {
-        // For now, the root AND should be there by construction. No
-        // code has been written that would eliminate the root AND term.
-        throw std::logic_error("Couldn't find root AND term");
-    }
-
-    AndTerm* incomingTerms = dynamic_cast<AndTerm*>(t.get());
-    if (incomingTerms) {
-        // Insert its elements then.
-        rootAnd->_terms.insert(rootAnd->_terms.begin(),
-                               incomingTerms->_terms.begin(),
-                               incomingTerms->_terms.end());
-    } else {
-        // Just insert the term as-is.
-        rootAnd->_terms.insert(rootAnd->_terms.begin(), t);
-    }
+    auto andTerm = std::make_shared<AndTerm>();
+    _rootOrTerm->addBoolTerm(andTerm);
+    return andTerm;
 }
 
 
 bool WhereClause::operator==(WhereClause& rhs) const {
```

Comment:
Normally the `rhs` parameter in this kind of the operators is always `const`. Otherwise a reader of the code expects side effects from this (by convention) read-only operator